src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +25,53 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result=std::monostate>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) { return Result(); }
   // Invoked after all the children of `module` is visited.
-  virtual void PostVisit(const Module& module) {}
+  virtual void PostVisit(const Module& module, Result& result) {}
+  virtual void FoldResult(const Module& module, Result child_result, Result& result) {}
   // Invoked before all the children of `block` is visited.
-  virtual void PreVisit(const Block& block) {}
+  virtual Result PreVisit(const Block& block) { return Result(); }
   // Invoked after all the children of `block` is visited.
-  virtual void PostVisit(const Block& block) {}
+  virtual void PostVisit(const Block& block, Result& result) {}
+  virtual void FoldResult(const Block& block, Result child_result, Result& result) {}
   // Invoked before all the children of `operation` is visited.
-  virtual void PreVisit(const Operation& operation) {}
+  virtual Result PreVisit(const Operation& operation) { return Result(); }
   // Invoked after all the children of `operation` is visited.
-  virtual void PostVisit(const Operation& operation) {}
+  virtual void PostVisit(const Operation& operation, Result& result) {}
+  virtual void FoldResult(const Operation& operation, Result child_result, Result& result) {}
 
-  void Visit(const Module& module) final override {
-    PreVisit(module);
+  Result Visit(const Module& module) final override {
+    Result result = PreVisit(module);
     for (const std::unique_ptr<Block>& block : module.blocks()) {
-      block->Accept(*this);
+      FoldResult(*block, block->Accept<Derived, Result>(*this), result);
     }
-    PostVisit(module);
+    PostVisit(module, result);
+    return result;
```

Comment:
If you're going to have `FoldResult`, I think it would make more sense to just have `FoldResult` return `Result`, similar to the other visitor methods. It's more in line with how the functional `fold` behaves and more in line with the [style guide recommendations](https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs). Also, you can use the `fold` utility we put in `//src/common/utils`.

Imagine you rewrote the above as follows:

```C++
Result Visit(const Module &module) final override {
  Result result = fold(module.blocks(), PreVisit(module), [&this](Result acc, const std::unique_ptr<Block> &block) {
    return FoldResult(*block, block->Accept<Derived, Result>(*this), std::move(acc));
  });
  PostVisit(module, result);
  return result;
}
```

Now `FoldResult` can just take `result` as a value:
```C++
virtual Result FoldResult(const Block &block, Result child_result, Result acc) {
  return acc;
}
```

You might also want to consider refactoring `PostVisit` to just return its result rather than taking a mutable reference parameter.

Reply:
This is originally how I had it. The issue I saw was that this requires moves of arbitrarily large result objects.
I'm happy to change it back though :)

Reply:
Ohh, actually I forgot that not everything *is* move-able. This further motivates applying the `child_result` to the `result` via side effects, but I'm still open to either option (a `unique_ptr` to the result value would be fine to avoid the moves).

Reply:
The thing about modern C++ is that it's good at avoiding copies mechanically, even if they appear to occur semantically. C++17 has copy elision, which basically says that if no live name refers to an object at the point when it is copied, the compiler is free to make the copy not occur. Note that I'm using an expansive definition of "name" here, including things like local variable names, field names, pointers to the value, the value being in a container, etc.

While we are stuck with C++'s legacy behavior and a very conservative notion of when a name is live, you can use `std::move` as a hint to the compiler to indicate that a name is no longer live. Using this judiciously can allow you to eliminate the lifetime concerns of passing around pointers while also not eating the performance penalty of a copy.

Reply:
Thanks! Done!